### PR TITLE
Bump exchange publish out metric when routing delayed messages

### LINKED ISSUE #157: No metrics for exchange publish out (count nor rate)
Tested in RabbitMQ version 3.8.12 Erlang 23.1 rabbitmq_delayed_message_exchange plugin 3.8.9.0199d11c

When publishing a delayed message to a delayed exchange no metrics for publish (out) is registered, neither when published nor routed later. Metric for publish (in) is there. 
Have checked that messages are being routed to a queue after the delay period.
Have tried with setting rates_mode to detailed but no difference. 

When publishing to a built-in exchange metrics are there.

### PR DESCRIPTION
**Pull request by gomoripeti**
fixes #157

## Proposed Changes

Bump the same metrics from the delayed exchanged process which are bumped from the channel process when routing a message without delay. This way we can get a sense of the rate messages leave the delayed exchange. Also incoming traffic for the queues that the delayed exchange routes to shows up in the queue graphs.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

I'd need some guidance on the GC of the core metrics.

I'm not sure whats the best way to test this as `rabbit_core_metrics` only has functions for writing but not reading and I'd avoid using functions from other plugins (like mgmt or prometheus). Only way I see currently is to read the ets table directly.

### CODE REVIEW

File: src/rabbit_delayed_message.erl [hunk 1]
```diff
@@ -260,3 +267,17 @@ recover_exchange_and_bindings(#exchange{name = XName} = X) ->
                               "recovered bindings for ~s",
                              [rabbit_misc:rs(XName)])
     end).
+
+bump_routed_stats(ExName, Qs) ->
+    rabbit_global_counters:messages_routed(amqp091, length(Qs)),
```
Comment (gomoripeti) [inline src/rabbit_delayed_message.erl:272]: is that correct that the Protocol is set to `amqp091`? at this point we don't know what the original protocol was.

File: src/rabbit_delayed_message.erl [hunk 2]
```diff
@@ -260,3 +267,17 @@ recover_exchange_and_bindings(#exchange{name = XName} = X) ->
...
+    [begin
+         QName = amqqueue:get_name(Q),
+         Key = {QName, ExName},
+         rabbit_core_metrics:channel_stats(queue_exchange_stats, publish, {self(), Key}, 1),
```
Comment (gomoripeti) [inline src/rabbit_delayed_message.erl:276]: would it cause any problem if the key contains the Pid of the delayed exchange process and not a channel process? Is this Pid used for something else than having a per-channel unique id?

File: src/rabbit_delayed_message.erl [hunk 3]
```diff
@@ -260,3 +267,17 @@ recover_exchange_and_bindings(#exchange{name = XName} = X) ->
...
+         %% Keys in the process dictionary are used to clean up the core metrics
+         %% FIXME see rabbit_channel:erase_queue_stats/1
+         %%put({queue_exchange_stats, Key}, none)
```
Comment (gomoripeti) [inline src/rabbit_delayed_message.erl:279]: should the delayed exchange process monitor all queues it ever published to and mark the metric when any of them is deleted?
Or would it cause problems if these metrics were never deleted?

File: src/rabbit_delayed_message.erl [hunk 4]
```diff
@@ -161,11 +161,18 @@ maybe_delay_first() ->
     end.
 
 route(#delay_key{exchange = Ex}, Deliveries) ->
+    ExName = Ex#exchange.name,
     lists:map(fun (#delay_entry{delivery = D}) ->
                       D2 = swap_delay_header(D),
                       Dests = rabbit_exchange:route(Ex, D2),
                       Qs = rabbit_amqqueue:lookup(Dests),
-                      rabbit_amqqueue:deliver(Qs, D2)
+                      case rabbit_queue_type:deliver(Qs, D2, stateless) of
```
Comment (gomoripeti) [inline src/rabbit_delayed_message.erl:169]: I'm not sure if it makes sense to check the result. If delivery to the first queue succeeds but fails for the second queue, the metrics will be incorrect anyway. I'm also not familiar with what stateless vs non-stateless delivery means.

**Comment by gomoripeti** [comment]
My motivation for this addition is that when a node is stopped and started after a while there could be a lot of messages already expired so it would be good to see the rate of traffic this results in towards the bound queues.

File: src/rabbit_delayed_message.erl [hunk 5]
```diff
@@ -161,11 +161,18 @@ maybe_delay_first() ->
...
+                          {ok, _QueueStates, _Actions} ->
+                              bump_routed_stats(ExName, Qs);
+                          {error, _} ->
+                              %% eg. stream coordinator unavailable
```
Comment (michaelklishin) [inline src/rabbit_delayed_message.erl:173]: This looks like a copy-and-paste artifact.

File: src/rabbit_delayed_message.erl (hunk 1 above)
Comment (michaelklishin) [reply src/rabbit_delayed_message.erl:272]: We don't but custom exchanges are rarely used via STOMP, so we can make this assumption.

File: src/rabbit_delayed_message.erl (hunk 4 above)
Comment (michaelklishin) [reply src/rabbit_delayed_message.erl:169]: This would be good enough for most users. Accurate metrics for events that happen in the future and reference things that might no longer exist (e.g. for error communication) is not something that most would expect.

File: src/rabbit_delayed_message.erl (hunk 2 above)
Comment (michaelklishin) [reply src/rabbit_delayed_message.erl:276]: This is just an ID, metric aggregators do not interact with these pids.

File: src/rabbit_delayed_message.erl (hunk 3 above)
Comment (michaelklishin) [reply src/rabbit_delayed_message.erl:279]: Monitoring every queue will be expensive. Keeping these samples around will leak
resources.

This exchange doesn't have access to the context. Since no other exchange is stateful and tries to do the work channels usually would, there was never any need to propagate the channel to `route/2`.

Comment (gomoripeti) [reply src/rabbit_delayed_message.erl:279]: Before I assumed that the metrics GC process uses the delete markers but now that I look at it, to me it looks like it diligently scans the whole table and removes all entries of deleted queues/exchanges it finds, including those that would be added by this PR (https://github.com/rabbitmq/rabbitmq-server/blob/master/deps/rabbit/src/rabbit_core_metrics_gc.erl#L191-L196) In this case there would be no metric leak and no need to monitor all queues. Do I see this correctly? What are metrics delete markers used for? I can't find usage in current master. Maybe a left-over from a previous metrics implementation?

Comment (michaelklishin) [reply src/rabbit_delayed_message.erl:279]: Ah, I forgot about that. It was added after this metrics design shipped, as a failsafe. Can you please verify that indeed leftover rows are always removed so this exchange can basically pass any value for an ID?

Also, if we pass a pid of a live process, would those rows be cleared? This is the question.

Comment (gomoripeti) [reply src/rabbit_delayed_message.erl]: did some tests of what happens when a delayed exchange or queue for which there are metrics is deleted

`rabbit_core_metics`, table: `channel_queue_exchange_metrics`
1. at delayed delivery a row with key "FakeChannelPid"+exchange+queue is written on the node of the exchange (where the original publish happened)
2. when a queue is deleted it calls `rabbit_core_metrics:queue_deleted/1` to mark all rows deleted for that queue on the node of the queue (might or might not be the same as publish node)
3. channels also monitor queues and call `rabbit_core_metrics:channel_queue_exchange_down/1` to mark entries for that queue as deleted (on the channel node) - this does not happen if the delay exchange does the publishing
4. `rabbit_mgmt_metrics_collector:aggregate_entry` will delete rows which are marked as deleted. This won't happen if a queue is deleted on another node than the publish.
5. there is a failsafe GC you mentioned that periodically goes through all entries and deletes (not just marks) those entries for which the queue or exchange does not exist or the Pid is not alive (this works for both channel or delay exchange PID). This makes sure there are no entries left behind even if 3./4. does not do its job. (This happens after a delay, not immediately when the resource is deleted.)

Additionally in case `rates_mode=detailed`
6. `rabbit_mgmt_metrics_collector:aggregate_entry` will call `insert_entry_ops(queue_exchange_stats_publish, ...)` if the core_metrics row is not marked for delete. However for a deleted exchange or queue this metric won't increase and the entries in `queue_exchange_stats_publish` table will expire/rotate out (I think this is done by `exometer_slide`)

> Also, if we pass a pid of a live process, would those rows be cleared? This is the question.
If the delayed exchange and the queue are still defined the row won't be cleared but I think this is expected. The `rabbit_delayed_message` gen_server process acts as a long running channel from this point of view.

So to sum up I couldnt find any leaking ETS row that would not be cleaned up after an exchange or queue is deleted.

**Comment by gomoripeti** [comment]
although initial testing is promising, please allow me to add some code comments and maybe a test case before merging

**Comment by michaelklishin** [comment]
Sure thing.

**Comment by michaelklishin** [comment]
Any updates on this?

**Comment by gomoripeti** [comment]
this is on the top of my todo list, but I was hindered by things like CodeBEAM etc :)

**Comment by gomoripeti** [comment]
this is ready for final review

**Review by michaelklishin — APPROVED** [review]

**Comment by michaelklishin** [comment]
Thank you!

**Comment by huang954333292** [comment]
Where is its link address?  “rabbitmq-delayed-message-exchange-3.10.25”  tell me ,please!!!

**Comment by huang954333292** [comment]
huang954333292@126.com

**Comment by michaelklishin** [comment]
@huang954333292 all releases of this plugin are available [from GitHub releases](https://github.com/rabbitmq/rabbitmq-delayed-message-exchange/releases). Usually you must use a release from the same minor version but not necessarily exactly the same patch release. Just use the most recent 3.10 patch release you see on the aforementioned releases page.

**Comment by huang954333292** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
